src/parser/command.rs

```diff
@@ -213,6 +217,48 @@ pub fn parse_defframe<'a>(input: ParserInput<'a>) -> ParserResult<'a, Instructio
     ))
 }
 
+/// Parse the contents of a `DEFGATE` instruction.
+pub fn parse_defgate<'a>(input: ParserInput<'a>) -> ParserResult<'a, Instruction> {
+    let (input, name) = token!(Identifier(v))(input)?;
+    let (input, parameters) = opt(delimited(
+        token!(LParenthesis),
+        separated_list1(token!(Comma), token!(Variable(v))),
+        token!(RParenthesis),
+    ))(input)?;
+    let (input, gate_type) = opt(preceded(
+        token!(As),
+        alt((
+            map(token!(Matrix), |()| GateType::Matrix),
+            map(token!(Permutation), |()| GateType::Permutation),
+        )),
+    ))(input)?;
+    let (input, _) = token!(Colon)(input)?;
+
+    let gate_type = gate_type.unwrap_or(GateType::Matrix);
+
+    let (input, matrix, permutation) = match gate_type {
+        GateType::Matrix => {
+            let (input, m) = parse_matrix(input)?;
+            (input, Some(m), None)
+        }
+        GateType::Permutation => {
+            let (input, p) = parse_permutation(input)?;
+            (input, None, Some(p))
+        }
+    };
+
+    Ok((
+        input,
+        Instruction::GateDefinition(GateDefinition {
+            name,
+            parameters: parameters.unwrap_or_default(),
+            matrix,
+            permutation,
+            r#type: gate_type,
+        }),
+    ))
+}
```

Comment:
With this change in place, what is your opinion on removing `r#type` on the `GateDefinition`?

The only place it is used is the instruction's `impl Display`, but we could get the same information by checking which variant of `GateSpecification` we have. At the moment, it is possible for `r#type` and `specification` to be out of sync, so I'm leaning towards removing `r#type` altogether.

Reply:
I agree, it's no longer needed as the information is now embedded in the enum variant.

Reply:
I think we still keep `GateType` here. While it isn't needed for `GateDefinition`, it's used as part of parsing a gate. We could remove it there too, but we'd have to either replace it with a string or have the parsing of type return a function to build the specification (instead of a flag that tells you what kind to build).

To me, both of those feel messier than just retaining `GateType`. Open to any other ideas though, or if someone thinks the approaches I lay out above are actually more idiomatic here.